GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -414,6 +429,52 @@ public virtual void GetFileTypeAndMode(string filePath, out FileType fileType, o
             }
         }
 
+        public virtual PathProjectionState GetPathProjectionState(string virtualPath)
+        {
+            this.GetChildNameAndParentKey(virtualPath, out string fileName, out string parentKey);
+            FolderEntryData data = this.GetProjectedFolderEntryData(
+                blobSizesConnection: null,
+                childName: fileName,
+                parentKey: parentKey);
+
+            if (data == null)
+            {
+                return PathProjectionState.NotFound;
+            }
+            else if (data.IsIncluded)
+            {
+                return PathProjectionState.Included;
+            }
+            else
+            {
+                return PathProjectionState.Excluded;
+            }
+        }
+
+        public bool TryAddIncludedFolder(string virtualPath)
+        {
+            try
+            {
+                this.GetChildNameAndParentKey(virtualPath, out string fileName, out string parentKey);
+                FolderEntryData data = this.GetProjectedFolderEntryData(
+                    blobSizesConnection: null,
+                    childName: fileName,
+                    parentKey: parentKey);
+
+                if (data != null && !data.IsIncluded)
+                {
+                    data.Include();
+                    this.includedFolderCollection.Add(virtualPath);

```

Comment:
Will this throw if the entry already exists?

Reply:
>Will this throw if the entry already exists?

No because the SQLite command is 
```
INSERT OR REPLACE INTO IncludedFolder (path) VALUES (@path);
```
Which will not do anything if it already exists since we are only using the primary key.